alicorn-expressions.lua

```diff
@@ -156,6 +154,9 @@ local function inferred_expression_pairhandler(env, a, b)
 
   -- combiner was an evaluated typed value, now it isn't
   local type_of_term, usage_count, term = infer(combiner, env.typechecking_context)
+	assert(type_of_term)
```

Comment:
Think we were avoiding `assert` due to a luajit performance issue, don't remember the details exactly. @aiverson ?

Reply:
Yes, assert should be avoided in hot loops because it causes a tracing abort, whereas `if cond then error(err) end` does not and can be jitted successfully on the trace where there is no error.